File: web/src/api/dispensing.ts
import { apiRequest } from "./client";

export interface DispenseRecord {
  id: string;
  prescription_id: string;
  prescription_item_id: string;
  patient_id: string;
  patient_name: string;
  product_id: string;
  product_name: string;
  location_id: string;
  quantity_dispensed: number;
  quantity_prescribed: number;
  pharmacist_id: string;
  pharmacist_name: string;
  technician_id: string;
  status: string;
  notes: string;
  witness_name: string;
  is_controlled: boolean;
  dispensed_at: string;
  created_at: string;
  updated_at: string;
}

export interface DispenseListResponse {
  data: DispenseRecord[];
  page: number;
  limit: number;
  total: number;
}

export interface PrescriptionItem {
  id: string;
  product_id: string;
  product_name: string;
  dosage: string;
  frequency: string;
  duration: string;
  quantity: number;
  refills_authorized: number;
  refills_used: number;
  notes: string;
}

export function listDispensing(page = 1, limit = 20, status = "", prescriptionId = "") {
  const params = new URLSearchParams({ page: String(page), limit: String(limit) });
  if (status) params.set("status", status);
  if (prescriptionId) params.set("prescription_id", prescriptionId);
  return apiRequest<DispenseListResponse>(`/dispensing?${params}`);
}

export function getDispenseRecord(id: string) {
  return apiRequest<DispenseRecord>(`/dispensing/${id}`);
}

export function createDispense(data: {
  prescription_item_id: string;
  quantity_dispensed: number;
  pharmacist_id: string;
  technician_id?: string;
  notes?: string;
  witness_name?: string;
  is_controlled?: boolean;
}) {
  return apiRequest<DispenseRecord>("/dispensing", { method: "POST", body: JSON.stringify(data) });
}

export function updateDispense(id: string, data: { technician_id?: string; witness_name?: string; notes?: string }) {
  return apiRequest<DispenseRecord>(`/dispensing/${id}`, { method: "PUT", body: JSON.stringify(data) });
}

export function updateDispenseStatus(id: string, status: string) {
  return apiRequest<DispenseRecord>(`/dispensing/${id}/status`, { method: "PUT", body: JSON.stringify({ status }) });
}

export function getLabelData(id: string) {
  return apiRequest<any>(`/dispensing/${id}/label`);
}

export function checkInteractions(productId: string, patientId: string) {
  return apiRequest<{ severity: string; message: string; interacting_product: string }[]>(`/drugs/${productId}/interactions?patient_id=${patientId}`);
}

export function checkAllergies(productId: string, patientId: string) {
  return apiRequest<{ severity: string; allergen: string; reaction: string }[]>(`/drugs/${productId}/allergy-check?patient_id=${patientId}`);
}

File: backend/models/dispensing.go
package models

import "time"

type DispenseRecord struct {
	ID                 string    `json:"id"`
	OrganisationID     string    `json:"organisation_id"`
	PrescriptionID     string    `json:"prescription_id"`
	PrescriptionItemID string    `json:"prescription_item_id"`
	PatientID          string    `json:"patient_id"`
	PatientName        string    `json:"patient_name"`
	ProductID          string    `json:"product_id"`
	ProductName        string    `json:"product_name"`
	LocationID         string    `json:"location_id"`
	QuantityDispensed  int       `json:"quantity_dispensed"`
	QuantityPrescribed int       `json:"quantity_prescribed"`
	PharmacistID       string    `json:"pharmacist_id"`
	PharmacistName     string    `json:"pharmacist_name"`
	TechnicianID       string    `json:"technician_id"`
	Status             string    `json:"status"`
	Notes              string    `json:"notes"`
	WitnessName        string    `json:"witness_name"`
	IsControlled       bool      `json:"is_controlled"`
	DispensedAt        string    `json:"dispensed_at"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

type DispenseLog struct {
	ID            string    `json:"id"`
	OrganisationID string   `json:"organisation_id"`
	DispenseID    string    `json:"dispense_id"`
	Action        string    `json:"action"`
	ActorID       string    `json:"actor_id"`
	Notes         string    `json:"notes"`
	CreatedAt     time.Time `json:"created_at"`
}
